### PR TITLE
Fix for external refunds synchronization

### PR DESCRIPTION
**Pull request by jnmrclmbsjse**
This PR includes:
1. Added condition if refunded and has payment.
2. Checking of existing refund refactor
3. Allow partial refund if previous refunds were below max refund amount

### CODE REVIEW

**Comment by jnmrclmbsjse** [comment]
Task linked: [CU-861mfzp8q Fix payment synchronization table](https://app.clickup.com/t/861mfzp8q)

File: src/StoreKeeper/WooCommerce/B2C/PaymentGateway/PaymentGateway.php [hunk 1]
```diff
@@ -350,18 +345,15 @@ public static function markRefundAsSynced($orderId, $skRefundId, $refundId): boo
     /**
      * @param $orderId
      * @param $refundId
-     *
-     * @return bool
      */
-    public static function refundExists($orderId, $refundId): ?bool
+    public static function refundExists($orderId, $refundId): ?int
```
Comment (ikulis) [inline src/StoreKeeper/WooCommerce/B2C/PaymentGateway/PaymentGateway.php:349]:
```suggestion
    public static function refundExists($orderId, $refundId): ?bool
```

Please retest the payments that one could not work correctly

File: src/StoreKeeper/WooCommerce/B2C/PaymentGateway/PaymentGateway.php [hunk 2]
```diff
@@ -370,11 +362,17 @@ public static function refundExists($orderId, $refundId): ?bool
...
         $results = $wpdb->get_results($sql, ARRAY_A);
-        if (!empty($results)) {
-            $payment_id = array_shift($results)['sk_refund_id'];
+
+        if (empty($results)) {
+            return false;
         }
 
-        return $payment_id;
+        $refund = array_shift($results);
```
Comment (ikulis) [inline src/StoreKeeper/WooCommerce/B2C/PaymentGateway/PaymentGateway.php:370]: why only first?

File: src/StoreKeeper/WooCommerce/B2C/PaymentGateway/PaymentGateway.php [hunk 3]
```diff
@@ -467,6 +465,28 @@ public function addGatewayClasses($default_gateway_classes)
...
+            $refundedValue = $order['paid_back_value_wt'];
+            $maxRefundValue = $paidValue - $refundedValue;
+            $floatEpsilon = 0.00001;
+            // Epsilon is only defined for PHP 7.2 and above
+            if (defined('PHP_FLOAT_EPSILON')) {
+                $floatEpsilon = PHP_FLOAT_EPSILON;
+            }
+
+            if ($maxRefundValue <= 0 || abs($refundAmount - $maxRefundValue) < $floatEpsilon) {
+                LoggerFactory::create('refund')->error('Order has refund on BackOffice already and refund is more than the expected amount', ['order_id' => $orderId, 'storekeeper_id' => $storeKeeperOrderId]);
```
Comment (ikulis) [inline src/StoreKeeper/WooCommerce/B2C/PaymentGateway/PaymentGateway.php:481]: maybe move the logger outside this function? than less params and checking function does not log errors

File: tests/unit/Endpoints/Webhooks/UpdateOrderTest.php [hunk 4]
```diff
@@ -91,35 +91,46 @@ public function dataProviderOrderRefund(): array
...
             'paidBackValueWt' => 50.00,
             'refundedPriceWt' => 50.00,
             'firstExpectedRefundCount' => 0,
             'secondExpectedRefundCount' => 0,
         ];
 
+        $data['backoffice order with partial refund'] = [
+            'paidValueWt' => 50.00,
+            'paidBackValueWt' => 25.00,
+            'refundedPriceWt' => 25.00,
```
Comment (ikulis) [inline tests/unit/Endpoints/Webhooks/UpdateOrderTest.php:112]:
```suggestion
            'paidBackValueWt' => 35.00,
            'refundedPriceWt' => 35.00,
```

makes better test to now split it in half

File: src/StoreKeeper/WooCommerce/B2C/PaymentGateway/PaymentGateway.php (hunk 2 above)
Comment (jnmrclmbsjse) [reply src/StoreKeeper/WooCommerce/B2C/PaymentGateway/PaymentGateway.php]: Changed this to check all refund records that matched order id and refund id.

File: src/StoreKeeper/WooCommerce/B2C/PaymentGateway/PaymentGateway.php (hunk 3 above)
Comment (jnmrclmbsjse) [reply src/StoreKeeper/WooCommerce/B2C/PaymentGateway/PaymentGateway.php]: Moved logger after method call.

**Pull request merged**